guides/howtos/test-using-mox.md

```diff
@@ -0,0 +1,88 @@
+# Test Using Mox with Tesla
+
+To mock HTTP requests in your tests using Mox with the Tesla HTTP client,
+follow these steps:
+
+## 1. Define a Mock Adapter
+
+First, define a mock adapter that implements the Tesla.Adapter behaviour. This
+adapter will intercept HTTP requests during testing.
+
+Create a file at `test/support/mocks.ex`:
+
+```elixir
+# test/support/mocks.ex
+Mox.defmock(MyApp.MockAdapter, for: Tesla.Adapter)
+```
+
+## 2. Configure the Mock Adapter for Tests
+
+In your `config/test.exs` file, configure Tesla to use the mock adapter you
+just defined:
+
+```elixir
+# config/test.exs
+config :tesla, adapter: MyApp.MockAdapter
+```
+
+If you are not using the global adapter configuration, ensure that your Tesla
+client modules are configured to use `MyApp.MockAdapter` during tests.
+
+## 3. Set Up Mocking in Your Tests
+
+Create a test module, for example `test/demo_test.exs`, and set up `Mox` to
+define expectations and verify them:
+
+```elixir
+defmodule MyApp.FeatureTest do
+  use ExUnit.Case, async: true
+
+  require Tesla.Test
+
+  setup context, do: Mox.set_mox_from_context(context)
+  setup context, do: Mox.verify_on_exit!(context)
+
+  test "example test" do
+    # Expect a single HTTP request to be made and return a JSON response
+    Tesla.Test.expect_tesla_call(
+      times: 1,
+      returns: Tesla.Test.json(%Tesla.Env{status: 200}, %{id: 1})
+    )
+
+    # Make the HTTP request using Tesla
+    # Mimic a use case where we create a user
+    assert :ok = create_user!(%{username: "johndoe"})
+
+    # Verify that the HTTP request was made and matches the expected parameters
+    Tesla.Test.assert_received_tesla_call(env, [])
+    Tesla.Test.assert_tesla_env(env, %Tesla.Env{
+      method: :post,
+      url: "https://acme.com/users",
+      body: %{username: "johndoe"},
+      status: 200,
+    })
```

Comment:
What's the advantage over standard pattern match `assert %Env{...} = ...`?

Reply:
Some convenience is based on previous experience of taking the given `env` **as is** and focusing on the matching alone, most of the time without having to unmarshal back the data in the JSON cases and excluding headers.

Reply:
I could imagine instead of:

```ex
Tesla.Test.assert_received_tesla_call(env, [])
Tesla.Test.assert_tesla_env(env, %Tesla.Env{
  method: :post,
  url: "https://acme.com/users",
  body: %{username: "johndoe"},
  status: 200,
})
```

doing something like this:

```ex
assert_received %Tesla.Env{
  method: :post,
  url: "https://acme.com/users",
  body: body,
  status: 200,
}

assert %{username: "johndoe"} = as_json(body)
```

This way you can use all the standard assertions with all their variants (`(assert|refute)(_receive)(d)`), extra params and test failure printout and it's clear what is really happening and what is being tested.

Reply:
That works as well; it is all about styling at that point, to be honest; 

![image](https://github.com/user-attachments/assets/02ed7969-24b7-4b09-96cd-5e91c67d31c1)

Reply:
showed the an extra example; we can always improve it as we go